From: Tomas Henzl <thenzl@redhat.com>
Date: Wed, 22 Apr 2009 15:18:58 +0300
Subject: [scsi] cciss: version change
Message-id: 49EF0B32.5080600@redhat.com
O-Subject: [RHEL5.4 PATCH4/4] cciss version change
Bugzilla: 474392

cciss version change

diff --git a/drivers/block/cciss.c b/drivers/block/cciss.c
index 19319d5..6deb939 100644
--- a/drivers/block/cciss.c
+++ b/drivers/block/cciss.c
@@ -52,16 +52,16 @@
 #include <linux/kthread.h>
 
 #define CCISS_DRIVER_VERSION(maj,min,submin) ((maj<<16)|(min<<8)|(submin))
-#define DRIVER_NAME "HP CISS Driver (v 3.6.20-RH2)"
+#define DRIVER_NAME "HP CISS Driver (v 3.6.20-RH3)"
 #define DRIVER_VERSION CCISS_DRIVER_VERSION(3,6,20)
 
 /* Embedded module documentation macros - see modules.h */
 MODULE_AUTHOR("Hewlett-Packard Company");
-MODULE_DESCRIPTION("Driver for HP Controller SA5xxx SA6xxx version 3.6.20-RH2");
+MODULE_DESCRIPTION("Driver for HP Controller SA5xxx SA6xxx version 3.6.20-RH3");
 MODULE_SUPPORTED_DEVICE("HP SA5i SA5i+ SA532 SA5300 SA5312 SA641 SA642 SA6400"
 			" SA6i P600 P800 P400 P400i E200 E200i E500 P700m"
 			" and HP Smart Array G2 SAS/SATA Controllers");
-MODULE_VERSION("3.6.20-RH2");
+MODULE_VERSION("3.6.20-RH3");
 MODULE_LICENSE("GPL");
 
 #include "cciss_cmd.h"
