Atomic failures aren't too interesting.

--- linux-2.6.15/include/linux/gfp.h~	2006-01-12 18:48:46.000000000 -0500
+++ linux-2.6.15/include/linux/gfp.h	2006-01-12 18:48:57.000000000 -0500
@@ -58,7 +58,7 @@ struct vm_area_struct;
 			__GFP_NOMEMALLOC|__GFP_HARDWALL)
 
 /* GFP_ATOMIC means both !wait (__GFP_WAIT not set) and use emergency pool */
-#define GFP_ATOMIC	(__GFP_HIGH)
+#define GFP_ATOMIC	(__GFP_HIGH | __GFP_NOWARN)
 #define GFP_NOIO	(__GFP_WAIT)
 #define GFP_NOFS	(__GFP_WAIT | __GFP_IO)
 #define GFP_KERNEL	(__GFP_WAIT | __GFP_IO | __GFP_FS)
