--- linux-2.6.18.i686.orig/Makefile	2007-03-21 13:54:07.000000000 -0400
+++ linux-2.6.18.i686/Makefile	2007-03-21 13:58:04.000000000 -0400
@@ -2,6 +2,8 @@
 PATCHLEVEL = 6
 SUBLEVEL = 18
 EXTRAVERSION = .4
+RHEL_MAJOR = 0
+RHEL_MINOR = 0
 NAME=Avast! A bilge rat!
 
 # *DOCUMENTATION*
@@ -864,7 +866,12 @@
 define filechk_version.h
 	(echo \#define LINUX_VERSION_CODE $(shell                             \
 	expr $(VERSION) \* 65536 + $(PATCHLEVEL) \* 256 + $(SUBLEVEL));     \
-	echo '#define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))';)
+        echo '#define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))'; \
+        echo \#define RHEL_MAJOR $(RHEL_MAJOR); \
+        echo \#define RHEL_MINOR $(RHEL_MINOR); \
+        echo \#define RHEL_RELEASE_CODE $(shell  \
+        expr $(RHEL_MAJOR) \* 256 + $(RHEL_MINOR));  \
+        echo '#define RHEL_RELEASE_VERSION(a,b) (((a) << 8) + (b))'; )
 endef
 
 include/linux/version.h: $(srctree)/Makefile FORCE


