Date: Tue, 10 Oct 2006 16:10:28 -0400 (EDT)
From: "Janice M. Girouard" <jgirouar@redhat.com>
Subject: Re: [RHEL5 PPC PATCH] RHBZ# 208938: RIT103210 - Got Call Trace
 message when remove veth module


Here it is inline:

--- a/drivers/net/ibmveth.c	2006-09-27 16:38:46.000000000 -0500
+++ b/drivers/net/ibmveth.c	2006-09-27 17:15:40.000000000 -0500
@@ -1131,7 +1131,9 @@ static void ibmveth_proc_register_adapte
 {
 	struct proc_dir_entry *entry;
 	if (ibmveth_proc_dir) {
-		entry = create_proc_entry(adapter->netdev->name, S_IFREG, ibmveth_proc_dir);
+		char u_addr[10];
+		sprintf(u_addr, "%x", adapter->vdev->unit_address);
+		entry = create_proc_entry(u_addr, S_IFREG, ibmveth_proc_dir);
 		if (!entry) {
 			ibmveth_error_printk("Cannot create adapter proc entry");
 		} else {
@@ -1146,7 +1148,9 @@ static void ibmveth_proc_register_adapte
 static void ibmveth_proc_unregister_adapter(struct ibmveth_adapter *adapter)
 {
 	if (ibmveth_proc_dir) {
-		remove_proc_entry(adapter->netdev->name, ibmveth_proc_dir);
+		char u_addr[10];
+		sprintf(u_addr, "%x", adapter->vdev->unit_address);
+		remove_proc_entry(u_addr, ibmveth_proc_dir);
 	}
 }

