--- ./fs/file.c.free	2011-09-14 13:58:38.000000000 +0400
+++ ./fs/file.c	2011-09-14 15:32:02.000000000 +0400
@@ -134,7 +134,9 @@ static void free_fdtable_rcu(struct rcu_
 		 */
 		return;
 	}
-	if (fdset_size <= PAGE_SIZE && fdarray_size <= PAGE_SIZE) {
+	if (!is_vmalloc_addr(fdt->open_fds) &&
+	    !is_vmalloc_addr(fdt->close_on_exec) &&
+	    !is_vmalloc_addr(fdt->fd)) {
 		kfree(fdt->open_fds);
 		kfree(fdt->close_on_exec);
 		kfree(fdt->fd);
